File: fix_fonts.py
import os

# Define replacement pairs
REPLACEMENTS = {
    # cart.jsp specific heavily corrupted lines
    "Gi?? hng c?a b?n ?ang tr?ng.": "Giỏ hàng của bạn đang trống.",
    "Xem ??n hng": "Xem đơn hàng",
    "Mua sch": "Mua sách",
    "V?? Kho Sch": "Về Kho Sách",
    "Thnh Ti??n": "Thành Tiền",
    "S? L??ng": "Số Lượng",
    "Xa Sch Ny?": "Xóa Sách Này?",
    "Xa": "Xóa",
    "Sch": "Sách",
    "Gi": "Giá",
    "Cn": "Còn",
    "cu?n": "cuốn",
    "T?ng": "Tổng",
    
    # Common ones
    "VN?": "VNĐ",
    "ch?n": "chọn",
    "Ch?n": "Chọn",
    "h?c": "học",
    "ngư?i": "người",
    "Ngư?i": "Người",
    "quy?n": "quyền",
    "Ti?n": "Tiền",
    "ti?n": "tiền",
    "th?i": "thời",
    "Th?i": "Thời",
    "? Trang chủ": "Về Trang chủ",
    "? Quay lại": "Quay lại",
    "? Duyệt sách": "Duyệt sách",
    "? My Library": "Về My Library",
    "??c": "Đọc",
    "?ể": "Để",
    "đư?ng": "đường",
    "Đư?ng": "Đường",
    "gi?": "giỏ",
    "Gi?": "Giỏ",
    "g?n": "gọn",
    "bộ l?c": "bộ lọc",
    "l?c nhanh": "lọc nhanh",
    "h?": "họ",
    "?nh": "Ánh" # Nguyễn Nhật Ánh
}

def fix_file(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except UnicodeDecodeError:
        return False
        
    original_content = content
    for bad_str, good_str in REPLACEMENTS.items():
        content = content.replace(bad_str, good_str)
        # Also handle standard ? instead of  if they decay differently
        bad_with_q = bad_str.replace("", "?")
        if bad_with_q != bad_str:
            content = content.replace(bad_with_q, good_str)
            
    if content != original_content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return True
    return False

def main():
    root_dir = "web/jsp"
    fixed_count = 0
    for dirpath, _, filenames in os.walk(root_dir):
        for filename in filenames:
            if filename.endswith(".jsp"):
                filepath = os.path.join(dirpath, filename)
                if fix_file(filepath):
                    print(f"Fixed: {filepath}")
                    fixed_count += 1
    print(f"Done. Fixed {fixed_count} files.")

if __name__ == "__main__":
    main()

File: FixFonts.java
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

public class FixFonts {
    private static int filesFixed = 0;
    private static final String R = "\uFFFD";
    
    private static final Map<String, String> REPLACEMENTS = new HashMap<>();
    
    static {
        REPLACEMENTS.put("Ti" + R + "?n phạt", "Tiền phạt");
        REPLACEMENTS.put("Ti" + R + "?n mặt", "Tiền mặt");
        REPLACEMENTS.put("Ti" + R + "?n (Amount)", "Tiền (Amount)");
        REPLACEMENTS.put("ngư" + R + "?i dùng", "người dùng");
        REPLACEMENTS.put("Ngư" + R + "?i nhận", "Người nhận");
        REPLACEMENTS.put("quy" + R + "?n", "quyền");
        REPLACEMENTS.put("V" + R + "? trang chủ", "Về trang chủ");
        REPLACEMENTS.put(R + "Về Trang chủ", "Về Trang chủ");
        REPLACEMENTS.put(R + "?" + R + "?c lúc", "Đọc lúc");
        REPLACEMENTS.put(R + "Để lần sau", "Để lần sau");
        REPLACEMENTS.put(R + "? My Library", "Về My Library");
        REPLACEMENTS.put(R + "? Trang", "Về Trang");
        REPLACEMENTS.put(R + "? Quay lại đơn", "Quay lại đơn");
        REPLACEMENTS.put(R + "?" + R + "?c sách", "Đọc sách");
        REPLACEMENTS.put(R + "?" + R + "?c ngay", "Đọc ngay");
        REPLACEMENTS.put("S" + R + "ch", "Sách");
        REPLACEMENTS.put("Giá" + R, "Giá");
        REPLACEMENTS.put("C" + R + "n", "Còn");
        REPLACEMENTS.put("X" + R + "a S" + R + "ch N" + R + "y?", "Xóa Sách Này?");
        REPLACEMENTS.put("X" + R + "a", "Xóa");
        REPLACEMENTS.put("VNĐ?", "VNĐ");
        REPLACEMENTS.put(R + "? Duyệt sách", "Duyệt sách");
        REPLACEMENTS.put("bộ l" + R + "?c", "bộ lọc");
        REPLACEMENTS.put("Nguyễn Nhật " + R + "?nh", "Nguyễn Nhật Ánh");
    }

    public static void main(String[] args) throws Exception {
        String rootDir = args.length > 0 ? args[0] : "c:\\Users\\tenma\\Downloads\\swp391-gr3-dung-merge-hao-borrow\\web\\jsp";
        System.out.println("Scanning: " + rootDir);
        try (Stream<Path> paths = Files.walk(new File(rootDir).toPath())) {
            paths.filter(Files::isRegularFile)
                 .filter(p -> p.toString().endsWith(".jsp"))
                 .forEach(path -> {
                     try { fixFile(path); }
                     catch (Exception e) { System.err.println("Error: " + path + " - " + e.getMessage()); }
                 });
        }
        System.out.println("Done. Files fixed: " + filesFixed);
    }

    private static void fixFile(Path path) throws Exception {
        String content = Files.readString(path, StandardCharsets.UTF_8);
        String original = content;

        for (Map.Entry<String, String> entry : REPLACEMENTS.entrySet()) {
            content = content.replace(entry.getKey(), entry.getValue());
        }

        if (!content.equals(original)) {
            Files.write(path, content.getBytes(StandardCharsets.UTF_8));
            filesFixed++;
            System.out.println("Fixed: " + path.getFileName());
        }
    }
}
